gengo/src/file_source/git.rs

```diff
@@ -107,23 +107,22 @@ impl<'repo> FileSource<'repo> for Git {
 
     fn overrides<O: AsRef<Path>>(&self, path: O) -> Overrides {
         let repo = self.repository.to_thread_local();
-        let state = {
-            let mut state = self.state.clone();
-            let Ok(platform) = state
-                .attr_stack
-                .at_path(path, Some(false), |id, buf| repo.objects.find_blob(id, buf))
+        let attr_matches = {
+            let mut attr_stack = self.state.attr_stack.clone();
```

Comment:
This line could be doing that, that's true. But what I see here still happily clones two structures in order to make them mutable. Also, `repo.to_thread_local()` is called once per path.

The problem is that the whole `self` seems to be tuned to be `Sync` and thus is read-only on top of that. The one coordinating the parallelism is responsible for passing in mutable state. I get it, each implementation has its own state, but if that's the case it must be a type parameter somewhere so it can be instantiated and passed in  mutably.

Maybe this happened due to `rayon` which clearly prefers `Sync` (share immutable state or shared mutable state behind locks), but even with `rayon` there are ways to get thread-local state that I *strongly* recommend getting into this method call.

This really is all I can say here, so: Happy Rusting :)!

Reply:
Alright, thanks again! I do have a habit of avoiding mutable references in favor of immutable references, which influenced my design decisions here. Thanks for the insight!